### ISSUE TITLE
Reduce parameters when creating a native-image

### ISSUE THREAD

**Issue by ilopmar** [micronaut-projects/micronaut-starter#299]
At this moment creating a native image for a Micronaut application requires a `native-image.properties` file with:

```
Args = -H:Name=basic-app \
       -H:Class=example.micronaut.Application
```

Which basically is the name of the binary (`basic-app` in this case) and the name of the `mainClass`.

Then, when creating a native image we do:
```
native-image --class-path build/libs/xxxxx.jar
```

As the fatjar is runnable (it contains the main class declared in the `MANIFEST.MF`) we could do the following:

- Simplify `native-image.properties` even more and keep only `Args = -H:Name=basic-app`
- Create the native image with `native-image -jar build/libs/xxxxx.jar`

I've tested it and it works.

The problem is that when creating an AWS app (with Controllers) we don't create a runnable fatjar and the main class is not a user class but a Micronaut one: `io.micronaut.function.aws.runtime.MicronautLambdaRuntime`, so for this use case we still need to define the `-H:Class=io.micronaut.function.aws.runtime.MicronautLambdaRuntime` option. This is probably related to https://github.com/micronaut-projects/micronaut-starter/issues/298

So, what do you think, is it worth it to do the changes to simplify even more the configuration for "regular" apps even if we need to still define the mainClass for AWS + GraalVM + Controllers?

**Comment by alvarosanchez** [comment]
Can the `-H:Name=basic-app` argument be passed directly? In which case, we could get rid of the properties file altogether

**Comment by graemerocher** [comment]
I think we maybe need to stop generating fat jars and this Graal config and move to doing it via build plugins.

This is what I did for the Gradle plugin you can run `./gradlew nativeImage` without the need for a `native-image.properties` file

**Comment by ilopmar** [comment]
@alvarosanchez yes, you can do `native-image -jar build/libs/basic-app-*-all.jar -H:Name=basic-app` but the idea is to put all the configuration in the `native-image.properties` so you don't need to pass those parameters (and remember them) when generating the native-image.

@graemerocher 

> I think we maybe need to stop generating fat jars and this Graal config and move to doing it via build plugins.

You mean fat jars for GraalVM or for everything?


Without the `native-image.properties` where do you define the name of the binary? In a configuration in the plugin? And the same if you need to pass a custom `--initialize-at-run-time`?
I personally think this is not a good approach because it makes everything more complex for users that know how GraalVM works. They need to learn how our plugin works instead of using the default configuration for GraalVM.

I think we should provide a `native-image.properties` for the users even if it's empty because we know the right directory and users may want/need to customize it. It doesn't hurt to have an empty file there but I see users frustrated when trying to add the file to the right directory for the first time.


I've always used the same example for explaining why I don't like and don't use this kind of plugins: Integrate Docker with Gradle. Using a Gradle DSL for defining the `Dockerfile` (FROM, RUN,...) in Gradle instead of using the standard `Dockerfile` that everyone knows it's not a good approach (for me at least). It's easier to provide a `Dockerfile` because again users know how it works, they can modify it as they want and they can run it from the command line. With an specific plugin they are coupled to that plugin.

**Comment by graemerocher** [comment]
for me personally what I would like to get a way with is that users have to add the `graalvm` feature in order to make a Micronaut app graalvm compatible.

This is achieved with the new Gradle plugin, with that in place it largely makes the `graalvm` feature redundant (except maybe for certain cases like lambdas)

Currently with the plugin we are able to infer the image name from the project name and the class to run from the configured main class:

https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/master/src/main/java/io/micronaut/gradle/graalvm/MicronautGraalPlugin.java#L44

If you need a custom argument you can do:

```
nativeImage {
      args "--initialize-at-run-time ..."
}
```

Maybe we can keep the `graalvm` feature around so that it adds a `native-image.properties`, but once the Gradle plugin is integrated (and the equivalent features in the Maven plugin) it becomes largely an optional feature.

**Comment by graemerocher** [comment]
Also regarding the docker file template, I have on the `docker-build` branch the ability to build docker images without the presence of a `Dockerfile` https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/docker-build/src/main/java/io/micronaut/gradle/graalvm/MicronautGraalPlugin.java

This is again makes the docker file an optional feature, but of course we can continue to define one because like you say it is good for users to be able to just use the Docker template and customize it.

However I do think we should be able to just build docker images out of the box with or without the present of a `Dockerfile`

**Comment by ilopmar** [comment]
If the plugin is flexible enough I agree we could remove pretty much all the GraalVM configuration. Once we publish a M1 release I will use it in a few test applications to see how they look like and provide more feedback.

Regarding Docker, I agree that it can be really good to be able to generate a docker images even if there is no `Dockerfile` present. But if we always provide a `Dockerfile` that will be used and I don't see users removing that file and letting Micronaut automatically manage it for them...

**Comment by graemerocher** [comment]
What I would like to see is that we provide out of the box solutions so you can get going with the minimal amount of infrastructure and templates, so the regardless if we do or do not have a `Dockerfile` I think we should provide that experience.

Another option is to add tasks to the plugins so you can do `./gradlew generateDockerFile` to generate the docker file for the sample template. This would be an extra step but certainly an option.

**Comment by alvarosanchez** [comment]
Fixed via #546
